<a href="https://colab.research.google.com/github/OptimalDecisions/sports-analytics-foundations/blob/main/pandas-basics/Pandas_Basics_2_3_Exploring_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Pandas Basics 2.3


In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
%matplotlib inline


In [4]:
url = "https://raw.githubusercontent.com/OptimalDecisions/sports-analytics-foundations/main/data/2022-2023%20NBA%20Player%20Stats%20-%20Regular.csv"
df = pd.read_csv(url, encoding = "ISO-8859-1", sep=";")



  # Examining, Describing & Summarizing Data

  <img src = "../img/sa_logo.png" width="100" align="left">

  Ram Narasimhan

  <br><br><br>

  << [Reading Data from files](Pandas_Basics_2_2_Reading_Files.ipynb)|[Examining, Describing & Summarizing Data](Pandas_Basics_2_3_Exploring_Data.ipynb) | [Column Operations](Pandas_Basics_2_4_Column_Operations.ipynb) >>



## Show (print) a dataframe

In [5]:
df

Rk            Player Pos  Age   Tm   G  GS    MP   FG   FGA  ...    FT%  \
0      1  Precious Achiuwa   C   23  TOR  55  12  20.7  3.6   7.3  ...  0.702   
1      2      Steven Adams   C   29  MEM  42  42  27.0  3.7   6.3  ...  0.364   
2      3       Bam Adebayo   C   25  MIA  75  75  34.6  8.0  14.9  ...  0.806   
3      4      Ochai Agbaji  SG   22  UTA  59  22  20.5  2.8   6.5  ...  0.812   
4      5      Santi Aldama  PF   22  MEM  77  20  21.8  3.2   6.8  ...  0.750   
..   ...               ...  ..  ...  ...  ..  ..   ...  ...   ...  ...    ...   
674  535    Thaddeus Young  PF   34  TOR  54   9  14.7  2.0   3.7  ...  0.692   
675  536        Trae Young  PG   24  ATL  73  73  34.8  8.2  19.0  ...  0.886   
676  537    Omer Yurtseven   C   24  MIA   9   0   9.2  1.8   3.0  ...  0.833   
677  538       Cody Zeller   C   30  MIA  15   2  14.5  2.5   3.9  ...  0.686   
678  539       Ivica Zubac   C   25  LAC  76  76  28.6  4.3   6.8  ...  0.697   

     ORB  DRB   TRB   AST  STL  BLK  TOV   PF   PTS  
0    1.8  4.1   6.0   0.9  0.6  0.5  1.1  1.9   9.2  
1    5.1  6.5  11.5   2.3  0.9  1.1  1.9  2.3   8.6  
2    2.5  6.7   9.2   3.2  1.2  0.8  2.5  2.8  20.4  
3    0.7  1.3   2.1   1.1  0.3  0.3  0.7  1.7   7.9  
4    1.1  3.7   4.8   1.3  0.6  0.6  0.8  1.9   9.0  
..   ...  ...   ...   ...  ...  ...  ...  ...   ...  
674  1.3  1.8   3.1   1.4  1.0  0.1  0.8  1.6   4.4  
675  0.8  2.2   3.0  10.2  1.1  0.1  4.1  1.4  26.2  
676  0.9  1.7   2.6   0.2  0.2  0.2  0.4  1.8   4.4  
677  1.7  2.6   4.3   0.7  0.2  0.3  0.9  2.2   6.5  
678  3.1  6.8   9.9   1.0  0.4  1.3  1.5  2.9  10.8  

[679 rows x 30 columns]

## Shape of a Dataframe

Typically, a dataframe will be a simple table of rows and columns. The`shape` method return a `tuple` containing the number of rows and the number of columns.

In [6]:
df.shape

(679, 30)

In [7]:
# prompt: print the number of rows of a df

num_rows = df.shape[0]
num_cols = df.shape[1]

print(f"This dataframe has {num_rows} rows and {num_cols} columns")

This dataframe has 679 rows and 30 columns


In [15]:
df.size # returns the total elements in the data frame (#rows times the number of columns)

20370

## df.info()

- info(): Provides a concise summary of the DataFrame, including the data types, non-null values, and memory usage.

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 679 entries, 0 to 678
Data columns (total 30 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Rk      679 non-null    int64  
 1   Player  679 non-null    object 
 2   Pos     679 non-null    object 
 3   Age     679 non-null    int64  
 4   Tm      679 non-null    object 
 5   G       679 non-null    int64  
 6   GS      679 non-null    int64  
 7   MP      679 non-null    float64
 8   FG      679 non-null    float64
 9   FGA     679 non-null    float64
 10  FG%     679 non-null    float64
 11  3P      679 non-null    float64
 12  3PA     679 non-null    float64
 13  3P%     679 non-null    float64
 14  2P      679 non-null    float64
 15  2PA     679 non-null    float64
 16  2P%     679 non-null    float64
 17  eFG%    679 non-null    float64
 18  FT      679 non-null    float64
 19  FTA     679 non-null    float64
 20  FT%     679 non-null    float64
 21  ORB     679 non-null    float64
 22  DR

## `head()` and `tail()`

- head(n): Returns the first n rows of the DataFrame. It's useful for a quick look at the beginning of the data.

- tail(n): Returns the last n rows of the DataFrame. It's useful for checking the end of the data.

In [9]:
df.head()

Rk            Player Pos  Age   Tm   G  GS    MP   FG   FGA  ...    FT%  \
0   1  Precious Achiuwa   C   23  TOR  55  12  20.7  3.6   7.3  ...  0.702   
1   2      Steven Adams   C   29  MEM  42  42  27.0  3.7   6.3  ...  0.364   
2   3       Bam Adebayo   C   25  MIA  75  75  34.6  8.0  14.9  ...  0.806   
3   4      Ochai Agbaji  SG   22  UTA  59  22  20.5  2.8   6.5  ...  0.812   
4   5      Santi Aldama  PF   22  MEM  77  20  21.8  3.2   6.8  ...  0.750   

   ORB  DRB   TRB  AST  STL  BLK  TOV   PF   PTS  
0  1.8  4.1   6.0  0.9  0.6  0.5  1.1  1.9   9.2  
1  5.1  6.5  11.5  2.3  0.9  1.1  1.9  2.3   8.6  
2  2.5  6.7   9.2  3.2  1.2  0.8  2.5  2.8  20.4  
3  0.7  1.3   2.1  1.1  0.3  0.3  0.7  1.7   7.9  
4  1.1  3.7   4.8  1.3  0.6  0.6  0.8  1.9   9.0  

[5 rows x 30 columns]

In [10]:
df.tail(3)

Rk          Player Pos  Age   Tm   G  GS    MP   FG  FGA  ...    FT%  \
676  537  Omer Yurtseven   C   24  MIA   9   0   9.2  1.8  3.0  ...  0.833   
677  538     Cody Zeller   C   30  MIA  15   2  14.5  2.5  3.9  ...  0.686   
678  539     Ivica Zubac   C   25  LAC  76  76  28.6  4.3  6.8  ...  0.697   

     ORB  DRB  TRB  AST  STL  BLK  TOV   PF   PTS  
676  0.9  1.7  2.6  0.2  0.2  0.2  0.4  1.8   4.4  
677  1.7  2.6  4.3  0.7  0.2  0.3  0.9  2.2   6.5  
678  3.1  6.8  9.9  1.0  0.4  1.3  1.5  2.9  10.8  

[3 rows x 30 columns]

In [16]:
df.sample(4)

Rk           Player Pos  Age   Tm   G  GS    MP   FG  FGA  ...    FT%  \
144  113       JD Davison  PG   20  BOS  12   0   5.5  0.7  1.6  ...  0.500   
557  434  Dennis Schröder  PG   29  LAL  66  50  30.1  4.1  9.8  ...  0.857   
453  358    Aaron Nesmith  SF   23  IND  73  60  24.9  3.5  8.1  ...  0.838   
384  301       Trey Lyles  PF   27  SAC  74   0  16.9  2.5  5.5  ...  0.815   

     ORB  DRB  TRB  AST  STL  BLK  TOV   PF   PTS  
144  0.2  0.6  0.8  0.9  0.2  0.2  0.3  0.4   1.6  
557  0.3  2.2  2.5  4.5  0.8  0.2  1.7  2.2  12.6  
453  0.8  2.9  3.8  1.3  0.8  0.5  1.0  3.2  10.1  
384  0.9  3.1  4.1  0.9  0.4  0.4  0.9  1.3   7.6  

[4 rows x 30 columns]

## List Columns

In [11]:
df.columns

Index(['Rk', 'Player', 'Pos', 'Age', 'Tm', 'G', 'GS', 'MP', 'FG', 'FGA', 'FG%',
       '3P', '3PA', '3P%', '2P', '2PA', '2P%', 'eFG%', 'FT', 'FTA', 'FT%',
       'ORB', 'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS'],
      dtype='object')

## Describe (a dataframe)

- `describe()`: Generates descriptive statistics of the numerical columns in the DataFrame, including count, mean, std deviation, min, 25%, 50%, 75%, and max.

In [12]:
df.describe()

Rk         Age           G          GS          MP          FG  \
count  679.000000  679.000000  679.000000  679.000000  679.000000  679.000000   
mean   265.976436   26.025037   43.338733   20.069219   19.463476    3.249190   
std    154.956296    4.325709   24.727306   25.766359    9.416316    2.362645   
min      1.000000   19.000000    1.000000    0.000000    1.000000    0.000000   
25%    132.500000   23.000000   22.000000    0.000000   12.150000    1.500000   
50%    264.000000   25.000000   45.000000    6.000000   18.800000    2.600000   
75%    399.500000   29.000000   65.500000   36.500000   27.550000    4.200000   
max    539.000000   42.000000   83.000000   83.000000   41.000000   11.200000   

              FGA         FG%          3P         3PA  ...         FT%  \
count  679.000000  679.000000  679.000000  679.000000  ...  679.000000   
mean     6.920913    0.462190    0.996465    2.782916  ...    0.711576   
std      4.795709    0.116682    0.862161    2.209427  ...    0.224774   
min      0.000000    0.000000    0.000000    0.000000  ...    0.000000   
25%      3.400000    0.415000    0.300000    1.000000  ...    0.667000   
50%      5.700000    0.454000    0.800000    2.400000  ...    0.760000   
75%      9.200000    0.505000    1.500000    4.150000  ...    0.841000   
max     22.200000    1.000000    4.900000   11.400000  ...    1.000000   

              ORB         DRB         TRB         AST         STL         BLK  \
count  679.000000  679.000000  679.000000  679.000000  679.000000  679.000000   
mean     0.842268    2.620177    3.461414    2.011635    0.601767    0.370398   
std      0.731867    1.715999    2.281317    1.891472    0.392063    0.367514   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.300000    1.400000    1.800000    0.800000    0.300000    0.100000   
50%      0.700000    2.300000    3.000000    1.300000    0.500000    0.300000   
75%      1.100000    3.400000    4.500000    2.700000    0.800000    0.500000   
max      5.100000    9.600000   12.500000   10.700000    3.000000    3.000000   

              TOV          PF         PTS  
count  679.000000  679.000000  679.000000  
mean     1.067305    1.660677    8.858763  
std      0.799478    0.770492    6.631226  
min      0.000000    0.000000    0.000000  
25%      0.500000    1.200000    4.100000  
50%      0.900000    1.600000    6.900000  
75%      1.400000    2.200000   11.550000  
max      4.100000    5.000000   33.100000  

[8 rows x 27 columns]

## Mean and Standar deviations (of columns)

In [14]:
df['Age'].mean(), df['Age'].std()

(26.02503681885125, 4.325709376933062)

From the above, we can see that the average age in the NBA is 26 years, and there is a standard deviation of 4.3 years

As we can see, Pandas provides us with several ways to "examine" a data frame.

In the next notebook, let's look at the many ways in which we can "operate" on columns of data.


<< [2.2 Reading Data from files](Pandas_Basics_2_2_Reading_Files.ipynb)|[2.3 Examining, Describing & Summarizing Data](Pandas_Basics_2_3_Exploring_Data.ipynb) | [2.4 Column Operations](Pandas_Basics_2_4_Column_Operations.ipynb) >>